packages/core/src/new-api/internal/execution/executionStateReducer.ts

```diff
@@ -0,0 +1,59 @@
+import { isDeploymentType } from "../../type-guards";
+import {
+  DeploymentExecutionState,
+  ExecutionState,
+  ExecutionStateMap,
+  ExecutionStatus,
+} from "../../types/execution-state";
+import { FutureStart, JournalableMessage } from "../../types/journal";
+
+export function executionStateReducer(
+  executionStateMap: ExecutionStateMap,
+  action: JournalableMessage
+) {
+  if (action.type === "execution-start") {
+    return {
+      ...executionStateMap,
+      [action.futureId]: initialiseExecutionStateFor(action),
+    };
+  }
+
+  if (action.type === "execution-success") {
+    const updatedExecutionState: DeploymentExecutionState = {
+      ...(executionStateMap[action.futureId] as DeploymentExecutionState),
```

Comment:
Why is this `as` required? I think it would be invalid without some extra checks.

Reply:
I have added an invariant that checks for deployment execution state.

For the moment static call, call and read event execution state are not covered, but there are issues to cover those todos.